test/aws-config/mzcompose.py

```diff
@@ -0,0 +1,282 @@
+
+
+def create_role(iam: IAMClient, effect: str, current_user: str) -> CreatedRole:
+    effect_name = effect.lower()
+    role_name = f"test-aws-config-{effect_name}-{SEED}"
+    policy_name = f"{effect_name}-{SEED}"
+    policy = gen_policy_document(effect)
+
+    create_response = iam.create_role(
+        RoleName=role_name,
+        AssumeRolePolicyDocument=gen_assume_role_document(current_user),
+    )
+    role_arn = create_response["Role"]["Arn"]
+    iam.put_role_policy(
+        RoleName=role_name,
+        PolicyName=policy_name,
+        PolicyDocument=policy,
+    )
+
+    return CreatedRole(name=role_name, arn=role_arn, policy_name=policy_name)
+
+
+def gen_policy_document(effect: str) -> str:
+    return """{
+   "Version":"2012-10-17",
+   "Statement":[
+      {
+         "Effect":"%(effect)s",
+         "Action":[
+            "s3:ListBucket",
+            "s3:GetBucketLocation"
+         ],
+         "Resource":"arn:aws:s3:::testdrive*"
+      },
+      {
+         "Effect":"%(effect)s",
+         "Action":[
+            "s3:GetObject",
+            "s3:GetObjectAcl"
+         ],
+         "Resource":"arn:aws:s3:::testdrive*/*"
+      }
+   ]
+}
+""" % {
+        "effect": effect
+    }
+
+
+def gen_assume_role_document(current_arn: str) -> str:
+    return """{
+    "Version": "2012-10-17",
+    "Statement": {
+        "Effect": "Allow",
+        "Principal": { "AWS": "%(current_arn)s" },
+        "Action": "sts:AssumeRole"
+    }
+}
+""" % {
+        "current_arn": current_arn
+    }
+
+
+def gen_profile_text(
+    session: boto3.Session, allowed_role_arn: str, denied_role_arn: str
+) -> str:
+    creds = session.get_credentials()
+    region = session.region_name
+
+    return f"""\
+[profile credentials]
+aws_access_key_id = {creds.access_key}
+aws_secret_access_key = {creds.secret_key}
+aws_session_token = {creds.token}
+region = {region}
+
+[profile allowed]
+source_profile = credentials
+role_arn = {allowed_role_arn}
+region = {region}
+
+# expected failure profiles
+
+[profile no-region]
+role_arn = {allowed_role_arn}
+aws_access_key_id = {creds.access_key}
+aws_secret_access_key = {creds.secret_key}
+aws_session_token = {creds.token}
+
+[profile denied]
+source_profile = credentials
+role_arn = {denied_role_arn}
+region = {region}
+
+[profile no-credentials]
+region = {region}
+
+# unexpected failure profiles
+
+# The region should be discovered in the source_profile, but it is not.
+[profile allowed-no-region]
+source_profile = credentials
+role_arn = {allowed_role_arn}
+"""
+
+
+def write_aws_config(c: Composition, text: str, service: str) -> None:
+    config = c.exec(service, "sh", "-c", "mkdir -p ~/.aws")
+    config = c.exec(
+        service, "sh", "-c", f"cat > ~/.aws/config <<EOF\n{text}EOF", quiet=True
+    )
+    config = c.exec(service, "sh", "-c", f"cat ~/.aws/config", capture=True)
+    if config.stdout != text:
+        raise UIError("Unable to verify that AWS config was correctly written")
```

Comment:
yeah that definitely would work, I specifically did this to avoid ever writing these credentials (which in general need to be admin credentials) to disk.

Reply:
I have removed the exec function and no longer exec into the container.